### PR TITLE
Dogstatsd/Statsd Mapper Feature

### PR DESCRIPTION
**Pull request by AlexandreYang**
## Description

Note: Depend on this PR https://github.com/DataDog/datadog-agent/pull/4575

Convert parts of statsd metric names to tags using mapping rules in `dogstatsd` using wildcard and regex patterns.

```yaml
dogstatsd_mapper_profiles:
  - name: <PROFILE_NAME>                    # e.g. "airflow", "consul", "some_database"
    prefix: <PROFILE_PREFIX>                # e.g. "airflow.", "consul.", "some_database."
    mappings:
      - match: <METRIC_TO_MATCH>                # e.g. `test.job.duration.*` to match `test.job.duration.my_job_name`
        match_type: <MATCH_TYPE>                # e.g. `wildcard` or `regex`
        name: <MAPPED_METRIC_NAME>              # e.g. `test.job.duration`
        tags:
          <TAG_KEY>: <TAG_VALUE_TO_EXPAND>      # e.g. `job_name: "$1"`, $1 is replaced by value capture by *
      - match: 'test.worker.*.*.start_time'     # to match `test.worker.<worker_type>.<worker_name>.start_time`
        name: 'test.worker.start_time'
        tags:
          worker_type: '$1'
          worker_name: '$2'
      - match: 'test\.task\.duration\.(\w+)\.(.*)'     # no need to escape in yaml context using single quote
        match_type: regex
        name: 'test.task'
        tags:
          task_type: '$1'
          task_name: '$2'
```

Using the mapping above we extract `job_type` and `job_name` from the metric name into tags before sending the metric to datadog backend.

The mappings is configured in `datadog.yaml` file. 

This feature will be very helpful for `Airflow integration` (see Airflow section for more info) and other integrations that expose mainly metrics via `statsd`.

If a incoming metric have already tags, no mapping is applied.

## Motivation

Statsd metrics frequently contain elements that should be tags instead.

For example `airflow.job.duration.<job_name>` could be `airflow.job.duration` with a `job_name` tag.

With a mapping we can make that conversion:
```
airflow.job.duration.<job_type>.<job_name>
=> airflow.job.duration + 2 tags (job_type, job_name)
```

### Wildcard Match Pattern

Match `dot-separated` metric names using wildcard, `*` as wildcard. Only characters defined by `allowedWildcardMatchPattern` (see code) are accepted.

Example:
```yaml
- match: 'airflow.job.duration.*'
   name: 'airflow.job.duration'
   tags:
     job_name: '$1'
```

### Regex Match Pattern

Match metric names using regex patterns.

Regex matching needed for cases where the captured part contain `.`, which happen for some metrics for Airflow. e.g. `dag_processing.last_run.seconds_ago.<dag_file>` the `<dag_file>` might contain `.`.

Example:
```yaml
- match: 'airflow\.job\.duration\.(\w+)\.(.+)'
   name: 'airflow.job.duration'
   tags:
     job_type: "$1"
     job_name: "$2"
```

### Mapped Metric Name

The `name` field is the name that the metric will be mapped to.  

Example:
```yaml
- match: 'airflow.job.duration.*'
   name: 'airflow.job.duration'
   tags:
     job_name: '$1'
```

You can use elements captured in the `name` field. 

Example:
```yaml
- match: 'airflow.test.*'
   name: 'airflow.$1.bar'
```

Using the previous example, if the incoming metric is `airflow.test.foo`, the final metric would be `airflow.foo.bar`

### Tags

Tags value can be expanded like `name` but not the tag key.

### Expansion of Captured Parts

The interpolation format is `$n` e.g. `$1`, `$2`, `$3`, ...

This format is also accepted:`${n}` e.g. `${1}`, `${2}`, `${3}`, ...

### Performance & Allocation

Most request should hit the LRU cache, the use of regex for the matching should not impact that much.

Once a metric to be matched is in the cache no further allocation is made.

The allocations are concentrated in two places:

- the mapping loaded from config
- the cache

**Benchmark** (see code)

```
BenchmarkMapperControl-8     2000000           743 ns/op
BenchmarkMapperControl-8     2000000           712 ns/op
BenchmarkMapperControl-8     2000000           671 ns/op

BenchmarkWithMapper-8        2000000           938 ns/op
BenchmarkWithMapper-8        2000000           906 ns/op
BenchmarkWithMapper-8        2000000           901 ns/op
```

Average 29% increase for `BenchmarkWithMapper` compared to `BenchmarkMapperControl`.

### Configuration via environment variable

It's currently not possible pass `dogstatsd_mapper_profiles` via env var, but if there is a use case later, we can add the support. For example, allowing user to pass `dogstatsd_mapper_profiles` as `json` object.

Similar to this: https://github.com/AlexandreYang/datadog-agent/blob/13e5494e4927d1dcc6789652e7f51205513cfa9d/pkg/trace/config/env_test.go#L220

### CODE REVIEW

File: pkg/dogstatsd/mapper/mapper.go
Comment (FlorianVeaux) [inline pkg/dogstatsd/mapper/mapper.go]: I couldn't be on the RFC you've presented but design question regarding performance here.
This whole logic is only applied if some mappings are defined in the config and only on metrics that are submitted without tags. This is great, couple questions though

1. Adding a mapper for just a few metrics (like airflow) will impact performance for any other statsd-type metric even though the dogstatsd service might already be heavily used. What do you think of branching out of this logic before? I expect that we'll have many mapper for each integration, could we imagine having a first matcher like "does my metric name contains 'airflow'" and if so apply the airflow mapper. Just an rough idea but having a way to limit the performance impact for the metric that don't require mappers.
2. You've run some benchmarks before but if my understanding is correct, the more mapping we add the more we expect performance to get worse (and in a linear way). That means that if 10 integration require 10 mappers each, we will check every single metric against all 100 mappers. Same rough idea as before, can we branch out before so that we just run the regex that are relevant?

All of that depends on one difficult aspect though, is it possible to differentiate metric that will require a mapper from those that don't without applying each pattern? There might be some easy things like having 'airflow' as the prefix, and can we rely on this in a general way.


Maybe something like this in the configuration and allowing `*` mappers that will be applied to every single metric?
```yaml
dogstatsd_mappings:
  'airflow':
    - match: <METRIC_TO_MATCH>
       match_type: glob
       name: <METRIC_NAME>
       ...
  'vault':
   - match: <METRIC_TO_MATCH>
      ...
  '*':
    - match: <METRIC_TO_MATCH>
       ....
```

File: releasenotes/notes/dogstatsd-mapper-668ce195a8c045b5.yaml
Comment (FlorianVeaux) [inline releasenotes/notes/dogstatsd-mapper-668ce195a8c045b5.yaml]: what do you think of keeping just `statsd` here?
```suggestion
    Add mapping feature to dogstatsd to convert parts of statsd
```

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: What about just `Map`. Usage `mapper.map("metric")` which sounds quite natural/intuitive.

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: Yes, make sense to avoid confusion. Renamed to wildcard.

File: pkg/config/config_template.yaml
Comment (AlexandreYang) [reply pkg/config/config_template.yaml]: Right. I just improved the doc.

File: releasenotes/notes/dogstatsd-mapper-668ce195a8c045b5.yaml
Comment (AlexandreYang) [reply releasenotes/notes/dogstatsd-mapper-668ce195a8c045b5.yaml]: Technically, it works for `dogstatsd` packets too, so just indicating `statsd` might be misleading.

File: pkg/dogstatsd/mapper/mapper.go
Comment (FlorianVeaux) [reply pkg/dogstatsd/mapper/mapper.go]: 👍

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: IIUC, your idea is to add a step to match the prefix before the actual matching.

Since we are using a LRU cache, each metric is only parsed once and then cached and the cache have a complexity of O(1). So, the performance should not decrease in the visible way since the mapping logic starting from `for _, mapping := range m.Mappings` is only executed once per metric.

Matching the prefix beforehand won't reduce much the performance but should significantly reduce the number of items cached if used before the cache retrieval (but might reduce a bit the perf since we need to check for prefix, then check in the cache).

Keeping the number of items in the cache low could be crucial since the cache key is a `statsd` metric name. The number of incoming entries can be high since `statsd` metric can contain elements like `job_name`, `task_id`, etc

So overall, I think it's a good thing to match the prefix. A little twist to your proposition. I would add a concept of `mapping_group` or maybe `profile` (like snmp), this will add flexibility in case we need other configuration later beside `prefix` at group level.

```
dogstatsd_mappings:
  # mapping_group 1
  - prefix: 'airflow.'
    mappings: 
      - match: <METRIC_TO_MATCH>
         match_type: glob
         name: <METRIC_NAME>
         ...
  # mapping_group 2
  - prefix: 'vault.'
    mappings: 
      - match: <METRIC_TO_MATCH>
         ...
      ...
  # mapping_group 3
  # match all if no prefix
  - mappings: 
      - match: <METRIC_TO_MATCH>
         ...
      ...

```

File: pkg/config/config_template.yaml
Comment (AlexandreYang) [reply pkg/config/config_template.yaml]: done

File: pkg/dogstatsd/server.go
Comment (arbll) [inline pkg/dogstatsd/server.go]: Not sure how problematic this is but this feature won't work well with containers. When the agent can identify the origin container it will [enrich the metric with tags automatically](https://github.com/DataDog/datadog-agent/blob/e42745bf55cf7f628b9e826b0528deef444b5530/pkg/dogstatsd/parser.go#L89-L95), breaking the assumption that no tags mean pure statsd metrics. Even if we removed the check we would not be able to cache the metrics by name since tags could change at any point when a container gets churned.

This identification happens in two cases:
1. When the metrics are sent over UDS
2. When the metrics are sent over UDP and contain a specific tag identifying the container

2\. is not possible in our case since statsd traffic can not contain tags. 1\. is however possible. While the traffic might come out of `airflow` & friends through UDP, it is common to [redirect the UDP traffic to UDS using socat](https://docs.datadoghq.com/developers/dogstatsd/unix_socket/?tab=docker#using-socat-as-a-proxy). Customers do this both to get better tagging but also because it's challenging to set up a UDP pipe between the agent and the application emitting statsd.

If it's doable it would be interesting to check if customers using airflow deploy it in a containerized version.

File: pkg/dogstatsd/server.go
Comment (arbll) [inline pkg/dogstatsd/server.go]: Nit: you could probably remove a few allocs by pre-building `sample.Tags` to the same size as `tags`. Probably not worth since I don't think we expect very high traffic through this path ?

File: pkg/dogstatsd/mapper/mapper.go
Comment (arbll) [inline pkg/dogstatsd/mapper/mapper.go]: As said in another comment IMO this should be moved to the config package. We should also map `MatchType` string to an enum there.

File: pkg/dogstatsd/mapper/mapper.go
Comment (arbll) [inline pkg/dogstatsd/mapper/mapper.go]: Nit: returning a pointer is more idiomatic and you convert it a few lines after this is called.

File: pkg/dogstatsd/mapper/mapper.go
Comment (arbll) [inline pkg/dogstatsd/mapper/mapper.go]: Nit: Create a struct for `name` and `tags`.

**Review by arbll — CHANGES_REQUESTED** [review]
Looks good, we need to discuss a few things around usage in containers

File: Gopkg.lock
Comment (AlexandreYang) [reply Gopkg.lock]: Sure, done.

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: Sure, done.

File: pkg/dogstatsd/server.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server.go]: As discussed, we might need to wait this PR https://github.com/DataDog/datadog-agent/pull/4427 that separate the pure parsing and the "tag enrichment" step.

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: Sure, done.

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: Sure, done.

File: Gopkg.lock
Comment (AlexandreYang) [reply Gopkg.lock]: Added an entry to Gopkg.toml, is that the right way?

File: pkg/dogstatsd/server.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server.go]: Let's wait #4427 :)

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: Sure, will move it to config package once the format is decided: https://github.com/DataDog/datadog-agent/pull/4471#discussion_r351188151

File: pkg/dogstatsd/server.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server.go]: It seems that most complex configs are kept within the packages they are intended to be used. e.g. `log`, `trace`:

- https://github.com/DataDog/datadog-agent/blob/09d20402c00f908c0fd56d6d42e214af214a3b7c/pkg/logs/config/processing_rules.go#L23-L31
- https://github.com/DataDog/datadog-agent/blob/18882ff18356f074dd86923bdb508efb7f2c7be1/pkg/trace/config/apply.go#L76-L91

This config object is both used for loading the config and then used in the processing logic, so IMHO it makes sense to keep it in the package where it has the most affinity, `doggstatsd`.

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: > As said in another comment IMO this should be moved to the config package.

The mapping configuration is quite specific to the mapper, so I think it's easier to maintain if it stays in the mapper package.

That's also what's done for `trace` https://github.com/DataDog/datadog-agent/blob/18882ff18356f074dd86923bdb508efb7f2c7be1/pkg/trace/config/apply.go

and logs https://github.com/DataDog/datadog-agent/blob/8979cb6aeec7e7e09d4c0b6dfa830936d380a13b/pkg/logs/config/endpoints.go

Maybe I can move those config objects in a config folder inside `dogstatsd`.

> We should also map MatchType string to an enum there.

I'm not sure it worth it here, since the MatchType is processed only once in `NewMetricMapper`

File: pkg/dogstatsd/server.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server.go]: PR update to use the mapper before the enrichment.

File: pkg/dogstatsd/server.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server.go]: That would make the code a bit harder to read. So, I would consider that as case of premature optimization that can be probably left for later if needed.

File: pkg/dogstatsd/server.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server.go]: Also answered here: https://github.com/DataDog/datadog-agent/pull/4471#discussion_r359456740

Comment (arbll) [reply pkg/dogstatsd/server.go]: Not sure I understand, both of your links point to a `config` package and seems to precisely deal with extracting config object in a separate package from the main logic ?
Example:
config package
https://github.com/DataDog/datadog-agent/blob/9ddbb7b0933ebb8f5002da2f867d2c425e5cd12d/pkg/logs/config/config.go#L54
logs package
https://github.com/DataDog/datadog-agent/blob/9c5a1a65ae7f571e2425e5a9edb7dd772820c515/pkg/logs/logs.go#L64

File: pkg/dogstatsd/mapper/mapper.go
Comment (arbll) [reply pkg/dogstatsd/mapper/mapper.go]: Let's discuss on the other thread

File: pkg/dogstatsd/server.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server.go]: @arbll So, should we like `logs` and `trace`, have a `config` package in `dogstatsd`: `datadog-agent/pkg/dogstatsd/config`

Maybe that's what you mean in the beginning.

I understood that you wanted me to mapper related config structures to `datadog-agent/pkg/config`.

Comment (arbll) [reply pkg/dogstatsd/server.go]: `datadog-agent/pkg/config` sounds fine to me, it's where the infra agent configuration is handled. 
To me it sounds logical that `trace/config` is called by the trace agent components, `logs/config` by the logs agent* components and `config` by the infra agent components.
Anyway, I don't want to block you on this, we can maybe ask people's opinion on slack?

*: I know it's not a separate process but historically it was and the code base reflects that

**Review by FlorianVeaux — DISMISSED** [review]
Your latest commits addressed my comment, looks great to me 👍 

I'll let @arbll confirm/comment as well

File: pkg/dogstatsd/server.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server.go]: @arbll Sure, if that's historically where we put configs for DSD, that's fine.

I just moved the config to the config package, let me know if that's ok for you.

File: pkg/dogstatsd/mapper/mapper.go
Comment (arbll) [inline pkg/dogstatsd/mapper/mapper.go]: giga nit: it's 2020 :yay:

**Review by arbll — COMMENTED** [review]
A few comments, should be mergable soon 👍

File: pkg/config/config.go
Comment (AlexandreYang) [reply pkg/config/config.go]: Seems not possible, but I'm not that familiar with viper:

```
=== RUN   TestMappingCases/Simple_OK_case
'dogstatsd_mapper_profiles: []'
Error: Could not parse dogstatsd_mapper_profiles: 1 error(s) decoding:
* '[0]' expected a map, got 'string'
```

It seems container is handling the `datadog.yaml` that way: https://github.com/DataDog/k8s-resources/blob/b1361cb1d7141660a9ff705c2a2243543adfe8f8/k8s/datadog-agent/templates/configmap.yaml#L45-L59

File: pkg/dogstatsd/server_bench_test.go
Comment (AlexandreYang) [reply pkg/dogstatsd/server_bench_test.go]: Sure, I will be glad to refactor to if the the method you linked is more precise.

Do you mind if I refactor once you have merged https://github.com/DataDog/datadog-agent/pull/4630 ? The benchmark code depend on some new code.

File: pkg/dogstatsd/enrich.go
Comment (AlexandreYang) [reply pkg/dogstatsd/enrich.go]: Make sense, refactored to make `Matched` private. Let me know if that's ok for you.

File: pkg/dogstatsd/mapper/mapper.go
Comment (AlexandreYang) [reply pkg/dogstatsd/mapper/mapper.go]: fixed.

**Comment by AlexandreYang** [comment]
@arbll I addressed/replied to the new feedbacks. Let me know if there is anything else I can do.

File: pkg/config/config.go
Comment (arbll) [reply pkg/config/config.go]: Yes internally we are mounting the configuration file in which case it would work without issues. That said mounting the file is not a requirement and probably not done in all setups. 

@mfpierre How do you deal with complex configuration options mapping to custom objects in env vars ? (Might be worth asking them directly as it will probably get lost in the notification spam)

File: pkg/dogstatsd/server_bench_test.go
Comment (arbll) [reply pkg/dogstatsd/server_bench_test.go]: It's fine to wait, just mentioning.

**Review by arbll — APPROVED** [review]

**Pull request merged**
